src/lidar_measurement_model_likelihood.cpp

```diff
@@ -0,0 +1,155 @@
+
+  double clip_z_min, clip_z_max;
+  pnh.param("clip_z_min", clip_z_min, -2.0);
+  pnh.param("clip_z_max", clip_z_max, 2.0);
+  clip_z_min_ = clip_z_min;
+  clip_z_max_ = clip_z_max;
+
+  double match_weight;
+  pnh.param("match_weight", match_weight, 5.0);
+  match_weight_ = match_weight;
+
+  double match_dist_min, match_dist_flat;
+  pnh.param("match_dist_min", match_dist_min, 0.2);
+  pnh.param("match_dist_flat", match_dist_flat, 0.05);
+  match_dist_min_ = match_dist_min;
+  match_dist_flat_ = match_dist_flat;
+}
+void LidarMeasurementModelLikelihood::setGlobalLocalizationStatus(
+    const size_t num_particles,
+    const size_t current_num_particles)
+{
+  if (current_num_particles <= num_particles)
+  {
+    num_points_ = num_points_default_;
+    return;
+  }
+  size_t num = num_points_default_ * num_particles / current_num_particles;
+  if (num < num_points_global_)
+    num = num_points_global_;
+
+  num_points_ = num;
+}
+
+typename pcl::PointCloud<LidarMeasurementModelBase::PointType>::Ptr
+LidarMeasurementModelLikelihood::filter(
+    const typename pcl::PointCloud<LidarMeasurementModelBase::PointType>::ConstPtr &pc) const
+{
+  const auto local_points_filter = [this](const LidarMeasurementModelBase::PointType &p)
+  {
+    if (p.x * p.x + p.y * p.y > clip_far_sq_)
+      return true;
+    if (p.x * p.x + p.y * p.y < clip_near_sq_)
+      return true;
+    if (p.z < clip_z_min_ || clip_z_max_ < p.z)
+      return true;
+    return false;
+  };
+  pcl::PointCloud<pcl::PointXYZI>::Ptr pc_filtered(new pcl::PointCloud<pcl::PointXYZI>);
+  *pc_filtered = *pc;
+  pc_filtered->erase(
+      std::remove_if(pc_filtered->begin(), pc_filtered->end(), local_points_filter), pc_filtered->end());
+  pc_filtered->width = 1;
+  pc_filtered->height = pc_filtered->points.size();
+
+  return sampler_.sample<LidarMeasurementModelBase::PointType>(pc_filtered, num_points_);
+}
+
+LidarMeasurementResult LidarMeasurementModelLikelihood::measure(
+    typename ChunkedKdtree<LidarMeasurementModelBase::PointType>::Ptr &kdtree,
+    const typename pcl::PointCloud<LidarMeasurementModelBase::PointType>::ConstPtr &pc,
+    const std::vector<Vec3> &origins,
+    const State6DOF &s) const
+{
+  if (!pc)
+    return LidarMeasurementResult(1, 0);
+  if (pc->size() == 0)
+    return LidarMeasurementResult(1, 0);
+  pcl::PointCloud<pcl::PointXYZI>::Ptr pc_particle(new pcl::PointCloud<pcl::PointXYZI>);
+  std::vector<int> id(1);
+  std::vector<float> sqdist(1);
+
+  float score_like = 0;
+  *pc_particle = *pc;
+  s.transform(*pc_particle);
+  size_t num = 0;
+  for (auto &p : pc_particle->points)
+  {
+    if (kdtree->radiusSearch(p, match_dist_min_, id, sqdist, 1))
+    {
+      const float dist = match_dist_min_ - std::max(sqrtf(sqdist[0]), match_dist_flat_);
+      if (dist < 0.0)
+        continue;
+
+      score_like += dist * match_weight_;
+      num++;
+    }
+  }
+  const float match_ratio = static_cast<float>(num) / pc_particle->points.size();
```

Comment:
put this into the argument below.

Reply:
I think naming it as `match_ratio` and passing it to the `LidarMeasurementResult()` is good at readability.
Do you have any reason?